### ISSUE TITLE
[SOLVED] How to cut/slice sample buffer

### ISSUE THREAD

**Issue by gotleg** [kcat/openal-soft#373]
First of all sorry if this is not the best place to ask questions.

I'm working on a sampler app written in C++/Qt/Openal.
App already allows to load and play multiple samples, change pitch but I feel stuck on the next feature.
I'm looking for the best way to cut / set the boundaries of a played sample to be able to use only a portion of a loaded sample.

At first I looked on the play part and I can't find any options on alSourcePlay to specify start / end point.
So I think I have to modify / make a modified copy of the buffer according to the boundaries I want to apply but I don't clearly see how the achieve this.

I already searched online for resources about this but can't find anything relevant (maybe I don't use right keywords). 

Any advices or ressources would be welcome :)

Thanks for your help.

**Comment by gotleg** [comment]
Just found part of a solution using  alSourcef(m_sourcesArray[sourceId], AL_SAMPLE_OFFSET, start);

**Comment by kcat** [comment]
There's currently not a way to limit what's used of a buffer. The best you can do is with `AL_SAMPLE_OFFSET`, but that only affects the initial start; it'll still play to the end of the buffer. I have thought about having some kind of 'sub-buffer' that only references part of another buffer, which would allow you to do what you want, but I'm not sure the best way to implement it.

The only way to get the behavior you want is to hold on to (or reload) the samples from the file and load the portion of samples you want into the desired buffer.

**Comment by gotleg** [comment]
Hi @kcat thanks for your help.

I finally found and implemented a solution very close to what you propose about sub-buffer.
All the buffer resizing is done with Qt before creating the openal buffer.

For those how could be interested here is my solution. I used WavFile and Utils class from Qt example  https://doc-snapshots.qt.io/qt5-5.11/qtmultimedia-multimedia-spectrum-example.html to avoid rewriting WavFile parsing.

Here are the main steps
- Scan samples directory and load samples using WavFile class (gives me QByteArray)
```
WavFile* wav = loadFile(filePath);
QByteArray buffer = wav->read(wav->size());
```

- I can now easily resize QByteArray buffer using these two methods
```
QByteArray resizedData = buffer.right(buffer.size() - startBytesToRemove);
resizedData.truncate(endBytesToRemove);
```

- Then I can create openal buffer with alBufferData
```
ALuint resizedBuffer;
alGenBuffers(1, &resizedBuffer);
alBufferData(resizedBuffer, AL_FORMAT_STEREO16, resizedData, resizedData.size(), 44100);
```

I hope this will help :)
Many thanks
